src/main/java/htsjdk/samtools/util/GZIIndex.java

```diff
@@ -0,0 +1,365 @@
+
+    /**
+     * Index entry mapping the block-offset (compressed offset) to the uncompressed offset where the
+     * block starts.
+     *
+     * @see <a href="http://github.com/samtools/htslib/issues/473">https://github.com/samtools/htslib/issues/473</a>
+     */
+    public static final class IndexEntry {
+        private final long compressedOffset;
+        private final long uncompressedOffset;
+
+        private IndexEntry(final long compressedOffset, final long uncompressedOffset) {
+            if (Long.signum(compressedOffset) == -1) {
+                throw new IllegalArgumentException("negative compressed offset: " + compressedOffset);
+            }
+            if (Long.signum(uncompressedOffset) == -1) {
+                throw new IllegalArgumentException("negative uncompressed offset: " + uncompressedOffset);
+            }
+            this.compressedOffset = compressedOffset;
+            this.uncompressedOffset = uncompressedOffset;
+        }
+
+        /** Returns the compressed offset (block-offset). */
+        public long getCompressedOffset() {
+            return compressedOffset;
+        }
+
+        /** Returns the uncompressed offset. */
+        public long getUncompressedOffset() {
+            return uncompressedOffset;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("IndexEntry={compressed=%d(0x%x),uncompressed=%d(0x%x)",
+                    compressedOffset, compressedOffset,
+                    uncompressedOffset, uncompressedOffset);
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            if (obj == null || !(obj instanceof IndexEntry)) {
+                return false;
+            }
+            final IndexEntry other = (IndexEntry) obj;
+            return compressedOffset == other.compressedOffset
+                    && uncompressedOffset == other.uncompressedOffset;
+        }
+
+        @Override
+        public int hashCode() {
+            return 31 * Long.hashCode(compressedOffset) + Long.hashCode(uncompressedOffset);
+        }
+    }
+
+    // first entry in the mapping (do not write down)
+    private static final IndexEntry FIRST_MAPPING = new IndexEntry(0, 0);
+
+    private final List<IndexEntry> entries;
+
+    // private constructors
+    private GZIIndex(final List<IndexEntry> entries) {
+        this.entries = entries;
+    }
+
+    /**
+     * Gets the number of blocks on the file.
+     *
+     * @return the number of blocks.
+     */
+    public int getNumberOfBlocks() {
+        return entries.size();
+    }
+
+    /**
+     * Gets an unmodifiable list with the index entries.
+     *
+     * @return index entries.
+     */
+    public List<IndexEntry> getIndexEntries() {
+        return Collections.unmodifiableList(entries);
+    }
+
+    /**
+     * Gets the virtual offset for seek with {@link BlockCompressedInputStream#seek(long)}.
+     *
```

Comment:
Could you give this a bit more detailed explanation in the comment?

Reply:
I added two paragraphs with more details.